pkg/urbit/noun/events.c

```diff
@@ -710,24 +673,31 @@ _ce_image_blit(u3e_image* img_u,
                c3_w*        ptr_w,
                c3_ws        stp_ws)
 {
-  c3_w i_w;
+  if ( 0 == img_u->pgs_w ) {
+    return;
+  }
+
+  c3_w siz_w = img_u->pgs_w * (1 << (u3a_page + 2));
 
   lseek(img_u->fid_i, 0, SEEK_SET);
-  for ( i_w=0; i_w < img_u->pgs_w; i_w++ ) {
-    if ( -1 == read(img_u->fid_i, ptr_w, (1 << (u3a_page + 2))) ) {
-      fprintf(stderr, "loom: image blit read: %s\r\n", strerror(errno));
-      c3_assert(0);
-    }
-#if 0
-    {
-      c3_w off_w = (ptr_w - u3_Loom);
-      c3_w pag_w = (off_w >> u3a_page);
+  if ( -1 == read(img_u->fid_i, ptr_w, siz_w) ) {
+    fprintf(stderr, "loom: image blit read: %s\r\n", strerror(errno));
+    c3_assert(0);
+  }
 
-      u3l_log("blit: page %d, mug %x\r\n", pag_w,
-              u3r_mug_words(ptr_w, (1 << u3a_page)));
+  {
+    c3_w i_w;
+    for ( i_w = 0; i_w < img_u->pgs_w; i_w++) {
+      c3_w pag_w = (u3a_outa(ptr_w) >> u3a_page) + i_w;
+      c3_w blk_w = pag_w >> 5;
+      c3_w bit_w = pag_w & 31;
+      u3P.dit_w[blk_w] &= ~(1 << bit_w);
     }
-#endif
-    ptr_w += stp_ws;
+  }
+
+  if ( 0 != mprotect(ptr_w, siz_w, PROT_READ) ) {
+    fprintf(stderr, "loom: live mprotect: %s\r\n", strerror(errno));
+    c3_assert(0);
```

Comment:
This change doesn't look right -- I suspect it only works because `south.bin` is alway a single page.

The old loop used to work a page at a time, iterating in opposite directions for north/south segments (`ptr_w += stp_ws;`). If I'm reading things correctly, `south.bin` is actually paginated left-to-right on disk, while it's right-to-left in memory. So I think you'll need to restore the loop and dirty/mprotect a page at a time.

Reply:
I overlooked this fact, so thanks for pointing it out. Out of curiosity, will it ever be possible to migrate to an approach where we iterate over the north and south segments in the same direction? From a performance standpoint, it would be ideal if we could issue a `read()` call and an  `mprotect()` call per segment rather than per page, which would require us to iterate from low addresses to high addresses. That said, my suspicion is that the current format of `south.bin` (pages at higher addresses coming first on disk) will make this infeasible if not impossible to switch.

Reply:
Yeah, it seems like that would be better. I'd be inclined to go with whatever approach will work best for demand paging. We'll have to migrate old snapshots if we change the layout, but that shouldn't be too difficult.

Reply:
I'm in favor of iterating over both segments in the same direction since it'll be easier to reason about, but I think the change should be put into a separate PR. For the sake of landing the inner road page tracking changes, I've switched back to a per-page approach to `read()` and `mprotect()` in `_ce_image_blit()`.